tests/federation_room_join_partial_state_test.go

```diff
@@ -3321,33 +3318,32 @@ func TestPartialStateJoin(t *testing.T) {
 		// sanity check - before the state has completed syncing state we would expect only one user
 		// to show up in the room list
 		res := terry.MustDoFunc(t, "GET", []string{"_matrix", "client", "v3", "publicRooms"})
-		body, err := ioutil.ReadAll(res.Body)
-		if err != nil {
-			t.Fatalf("something broke: %v", err)
-		}
-		numJoinedMembers := gjson.GetBytes(body, "chunk.0.num_joined_members")
-		if numJoinedMembers.Int() != 1 {
-			t.Fatalf("Expected 1 users, returned %s", numJoinedMembers)
-		}
+
+		must.MatchResponse(t, res, match.HTTPResponse{
+			StatusCode: 200,
+			JSON: []match.JSON{
+				match.JSONKeyEqual("chunk.0.num_joined_members", 1),
+			}})
 
 		// finish syncing the state
 		psjResult.FinishStateRequest()
-		terry.MustSyncUntil(t,
-			client.SyncReq{},
-			client.SyncJoinedTo(terry.UserID, psjResult.ServerRoom.RoomID),
-		)
-
-		// check the number of joined users, it should now be 3
-		res2 := terry.MustDoFunc(t, "GET", []string{"_matrix", "client", "v3", "publicRooms"})
-		body2, err2 := ioutil.ReadAll(res2.Body)
-		if err2 != nil {
-			t.Fatalf("something broke: %v", err2)
-		}
-		numJoinedMembers2 := gjson.GetBytes(body2, "chunk.0.num_joined_members")
-		if numJoinedMembers2.Int() != 3 {
-			t.Fatalf("Expected 3 users, returned %s", numJoinedMembers2)
-		}
-
+		awaitPartialStateJoinCompletion(t, psjResult.ServerRoom, terry)
+
+		// the rooms stats are updated by a background job which is not guaranteed to have completed by the time
+		// the state sync has completed. We check for up to 3 seconds that the job has completed. The number of joined
+		// users should now be 3: one local user (terry) and two remote (charlie and derek)
```

Comment:
The background job is a synapse implementation detail. I'd either clarify that this refers to synapse or just state that there's no guarantee, without referring to the background job.

Reply:
I added info about the background job being Synapse, hopefully it's clearer.